67_SubAllocatedDescriptorSet/main.cpp

```diff
@@ -0,0 +1,290 @@
+// Copyright (C) 2018-2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+
+#include "nbl/video/alloc/SubAllocatedDescriptorSet.h"
+
+#include "../common/BasicMultiQueueApplication.hpp"
+#include "../common/MonoAssetManagerAndBuiltinResourceApplication.hpp"
+
+#include "nbl/builtin/hlsl/random/xoroshiro.hlsl"
+
+using namespace nbl;
+using namespace core;
+using namespace system;
+using namespace ui;
+using namespace asset;
+using namespace video;
+
+class SubAllocatedDescriptorSetApp final : public examples::MonoDeviceApplication, public examples::MonoAssetManagerAndBuiltinResourceApplication
+{
+		using device_base_t = examples::MonoDeviceApplication;
+		using asset_base_t = examples::MonoAssetManagerAndBuiltinResourceApplication;
+
+		smart_refctd_ptr<nbl::video::ICommandPoolCache> m_poolCache;
+		smart_refctd_ptr<nbl::video::SubAllocatedDescriptorSet> m_subAllocDescriptorSet;
+
+		smart_refctd_ptr<ISemaphore> m_timeline;
+		uint64_t m_iteration = 0;
+		constexpr static inline uint64_t MaxIterations = 200;
+		constexpr static inline uint64_t MaxDescriptors = 512;
+		constexpr static inline uint64_t MaxAllocPerFrame = 10;
+		constexpr static uint32_t AllocatedBinding = 0;
+		smart_refctd_ptr<IGPUImageView> m_descriptorImages[MaxDescriptors];
+		smart_refctd_ptr<IGPUBuffer> m_descriptorBuffers[MaxDescriptors];
```

Comment:
declaring them like this might keep them alive via shared ownership, an it might hide lifetime tracking issues.

Reply:
alternative is to keep a `container<pair<binding,offset>>` of everything you haven't freed (stil occcupant) where the container can be a vector or a set

then you draw a random number of elements from that to multi_deallocate